flink-runtime/src/main/java/org/apache/flink/runtime/security/token/KerberosDelegationTokenManager.java

```diff
@@ -110,13 +139,62 @@ public void obtainDelegationTokens(Credentials credentials) {
      * task managers.
      */
     @Override
-    public void start() {
-        LOG.info("Starting renewal task");
+    public void start() throws Exception {
+        checkNotNull(scheduledExecutor, "Scheduled executor must not be null");
+        checkNotNull(executorService, "Executor service must not be null");
+        checkState(tgtRenewalFuture == null, "Manager is already started");
+
+        if (!kerberosRenewalPossibleProvider.isRenewalPossible()) {
+            LOG.info("Renewal is NOT possible, skipping to start renewal task");
+            return;
+        }
+
+        startTGTRenewal();
+    }
+
+    private void startTGTRenewal() throws IOException {
+        LOG.debug("Starting credential renewal task");
+
+        UserGroupInformation currentUser = UserGroupInformation.getCurrentUser();
+        if (currentUser.isFromKeytab()) {
+            // In Hadoop 2.x, renewal of the keytab-based login seems to be automatic, but in Hadoop
+            // 3.x, it is configurable (see hadoop.kerberos.keytab.login.autorenewal.enabled, added
+            // in HADOOP-9567). This task will make sure that the user stays logged in regardless of
+            // that configuration's value. Note that checkTGTAndReloginFromKeytab() is a no-op if
+            // the TGT does not need to be renewed yet.
+            long tgtRenewalPeriod = configuration.get(KERBEROS_RELOGIN_PERIOD).toMillis();
+            tgtRenewalFuture =
+                    scheduledExecutor.scheduleAtFixedRate(
+                            () ->
+                                    executorService.execute(
+                                            () -> {
+                                                try {
+                                                    LOG.debug("Renewing TGT");
+                                                    currentUser.checkTGTAndReloginFromKeytab();
+                                                    LOG.debug("TGT renewed successfully");
+                                                } catch (Exception e) {
+                                                    LOG.error("Error while renewing TGT", e);
+                                                }
+                                            }),
+                            tgtRenewalPeriod,
+                            tgtRenewalPeriod,
+                            TimeUnit.MILLISECONDS);
+            LOG.debug("Credential renewal task started and reoccur in {} ms", tgtRenewalPeriod);
```

Comment:
There is some inconsistency in the log messages, there is a mix of references to TGT and credentials, these should be unified. Also it would be great to add a reference to the class javadoc about what the TGT really stands for.

Reply:
Changed to TGT all the places. In Kerberos there are hundreds of cloudy things and adding them to classes and/or functions doesn't help. My intention is to add a complete doc like [this](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/deploy/security/README.md) which includes every conceptual/architectural explanation. Here we update Ticket Granting Ticket and there is a description why needed:
```
            // In Hadoop 2.x, renewal of the keytab-based login seems to be automatic, but in Hadoop
            // 3.x, it is configurable (see hadoop.kerberos.keytab.login.autorenewal.enabled, added
            // in HADOOP-9567). This task will make sure that the user stays logged in regardless of
            // that configuration's value. Note that checkTGTAndReloginFromKeytab() is a no-op if
            // the TGT does not need to be renewed yet.
```